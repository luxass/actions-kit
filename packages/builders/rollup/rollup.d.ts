declare module "actions-kit/config" {
	import type { RollupOptions } from "rollup";

	interface ActionsKitConfig {
		rollup?: RollupOptions;
	}
}

export {};
