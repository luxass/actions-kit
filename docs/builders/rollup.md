# Rollup Builder

Build GitHub Actions using Rollup.

## Installation

::: code-group

```bash [pnpm]
pnpm add -D actions-kit @actions-sdk/rollup-builder
```

```bash [yarn]
yarn add -D actions-kit @actions-sdk/rollup-builder
```

```bash [npm]
npm install -D actions-kit @actions-sdk/rollup-builder
```

```bash [bun]
bun install -D actions-kit @actions-sdk/rollup-builder
```

:::

Set the `builder` in your `actions-kit.config.ts` file:

```ts [actions-kit.config.ts]
import { defineConfig } from "actions-kit/config";

export default defineConfig({
  builder: "rollup",
});
```

> [!TIP]
> If you are using TypeScript, you will have to either add `@actions-sdk/rollup-builder/types` to `compilerOptions.types` or add a triple slash directive to the top of your file.

```ts
/// <reference types="@actions-sdk/rollup-builder/types" />
```

### Customizing Rollup

You can customize the Rollup options by adding an `rollup` key to your `actions-kit.config.ts` file:

```ts [actions-kit.config.ts]
import { defineConfig } from "actions-kit/config";

export default defineConfig({
  builder: "rollup",
  rollup: {},
});
```
