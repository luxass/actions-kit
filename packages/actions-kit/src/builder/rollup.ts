import { inferModuleType, inferOutputFilename } from "../utils";
import type { Config } from "../config";
import actionsKit from "unplugin-actions-kit/rollup";
import { join } from "node:path";
import { builtinModules } from "node:module";

export async function build(config: Config) {
	const rollup = await import("rollup").then((m) => m.rollup);
	const outputFileName = await inferOutputFilename(config);
	const libraryType = await inferModuleType(config, outputFileName);

	const rollupConfig = config.rollup;

	const bundle = await rollup({
		input: "./src/index.ts",
		external: [...builtinModules, ...builtinModules.map((m) => `node:${m}`)],
		output: {
			file: join(process.cwd(), "dist", outputFileName),
			format: libraryType,
			exports: "auto",
		},
		plugins: [
			actionsKit({
				// TODO: allow users to specify it.
				actionPath: join(process.cwd(), "./action.yml"),
				inject: true,
			}),
		],
	});

	const wrote = await bundle.write({
		format: libraryType,
		sourcemap: false,
	});

  console.log(wrote);
}
