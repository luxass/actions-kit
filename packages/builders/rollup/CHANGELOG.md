# @actions-sdk/rollup-builder

## 0.2.1

### Patch Changes

- [#23](https://github.com/luxass/actions-kit/pull/23) [`46032ce`](https://github.com/luxass/actions-kit/commit/46032ce171655d8a2eb7277238664aa386485b3b) Thanks [@luxass](https://github.com/luxass)! - fix package exports and add missing cli entrypoint for `actions-kit`

- Updated dependencies [[`46032ce`](https://github.com/luxass/actions-kit/commit/46032ce171655d8a2eb7277238664aa386485b3b)]:
  - actions-kit@0.2.2

## 0.2.0

### Minor Changes

- [`bcf5300`](https://github.com/luxass/actions-kit/commit/bcf53008a191840aee634e040f4d9c53cfbf2a8b) Thanks [@luxass](https://github.com/luxass)! - change how builders work, for more information https://actions-kit.pages.dev/builders/
