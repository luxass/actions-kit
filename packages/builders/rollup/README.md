# @actions-sdk/rollup-builder

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]

Rollup builder for Actions Kit.

## Installation

```bash
pnpm add -D actions-kit @actions-sdk/rollup-builder
```

> [!NOTE]
> Documentation is available [here](https://actions-kit.pages.dev/builders/rollup)

## ðŸ“„ License

Published under [MIT License](./LICENSE).

[npm-version-src]: https://img.shields.io/npm/v/@actions-sdk/rollup-builder?style=flat&colorA=18181B&colorB=4169E1
[npm-version-href]: https://npmjs.com/package/@actions-sdk/rollup-builder
[npm-downloads-src]: https://img.shields.io/npm/dm/@actions-sdk/rollup-builder?style=flat&colorA=18181B&colorB=4169E1
[npm-downloads-href]: https://npmjs.com/package/@actions-sdk/rollup-builder
