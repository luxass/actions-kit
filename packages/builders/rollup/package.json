{
  "name": "@actions-sdk/rollup-builder",
  "version": "0.2.7",
  "description": "Rollup builder for Actions Kit",
  "type": "module",
  "author": {
    "name": "Lucas Nørgård",
    "email": "lucasnrgaard@gmail.com",
    "url": "https://luxass.dev"
  },
  "packageManager": "pnpm@10.14.0",
  "license": "MIT",
  "homepage": "https://github.com/luxass/actions-kit",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/luxass/actions-kit.git",
    "directory": "packages/builders/rollup"
  },
  "bugs": {
    "url": "https://github.com/luxass/actions-kit/issues"
  },
  "keywords": [
    "rollup",
    "actions-kit",
    "github actions",
    "actions"
  ],
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.cts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsdown --tsconfig=./tsconfig.build.json",
    "dev": "tsdown --watch",
    "clean": "git clean -xdf dist node_modules",
    "lint": "eslint .",
    "typecheck": "tsc --noEmit"
  },
  "peerDependencies": {
    "@rollup/plugin-commonjs": "catalog:builders",
    "@rollup/plugin-node-resolve": "catalog:builders",
    "@rollup/plugin-typescript": "catalog:builders",
    "actions-kit": "workspace:*",
    "rollup": "catalog:builders"
  },
  "dependencies": {
    "defu": "catalog:prod",
    "unplugin-actions-kit": "workspace:*"
  },
  "devDependencies": {
    "@actions-kit/tsconfig": "workspace:*",
    "@actions-kit/tsdown-config": "workspace:*",
    "@luxass/eslint-config": "catalog:linting",
    "@rollup/plugin-commonjs": "catalog:builders",
    "@rollup/plugin-node-resolve": "catalog:builders",
    "@rollup/plugin-typescript": "catalog:builders",
    "@types/node": "catalog:types",
    "actions-kit": "workspace:*",
    "eslint": "catalog:linting",
    "publint": "catalog:dev",
    "rollup": "catalog:builders",
    "tsdown": "catalog:dev",
    "typescript": "catalog:dev",
    "vitest": "catalog:testing"
  }
}
