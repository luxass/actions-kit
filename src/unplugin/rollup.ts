/**
 * This entry file is for Rollup plugin.
 *
 * @module
 */

import { createRollupPlugin } from "unplugin";
import type unplugin from "./";
import { unpluginFactory } from "./";

/**
 * Rollup plugin
 *
 * @example
 * ```ts
 * // rollup.config.js
 * import actionsKit from "actions-kit/unplugin//rollup"
 *
 * export default {
 *   plugins: [actionsKit()],
 * }
 * ```
 */
export default createRollupPlugin(unpluginFactory) as typeof unplugin.rollup;
