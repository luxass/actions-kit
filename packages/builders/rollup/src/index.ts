import { rollup } from "rollup";
import { defineBuilder, type BuildOutput } from "actions-kit/builder";
import { inferModuleType, inferOutput } from "actions-kit/builder-utils";
import { defu } from "defu";
import { stat } from "node:fs/promises";
import { builtinModules } from "node:module";
import { join } from "node:path";
import type { RollupOptions } from "rollup";
import RollupActionsKit from "unplugin-actions-kit/rollup";

export default function rollupBuilder(options: RollupOptions = {}) {
	return defineBuilder({
		name: "rollup",
		build: async ({ cwd, config}) => {
			const { filename, dir } = await inferOutput(config);
			const libraryType = await inferModuleType(config, filename);

			const [commonjs, resolve, typescript] = await Promise.all([
				import("@rollup/plugin-commonjs").then((m) => m.default),
				import("@rollup/plugin-node-resolve").then((m) => m.default),
				import("@rollup/plugin-typescript").then((m) => m.default),
			]);

			const rollupOptions = defu(options, {
				input: config.build?.input || join(cwd, "src/index.ts"),
				external: [...builtinModules, ...builtinModules.map((m) => `node:${m}`)],
				output: {
					file: join(cwd, dir, filename),
					format: libraryType,
					exports: "auto",
				},
				plugins: [
					typescript({
						tsconfig: "./tsconfig.json",
						declaration: false,
					}),
					resolve({
						preferBuiltins: true,
					}),
					commonjs(),
					RollupActionsKit({
						// TODO: allow users to specify it.
						actionPath: join(cwd, "./action.yml"),
						inject: config.inject,
						autocomplete: config.autocomplete,
					}),
				],
				onwarn: (warning) => {
					if (
						warning.code === "UNRESOLVED_IMPORT" ||
						warning.code === "CIRCULAR_DEPENDENCY" ||
						warning.code === "EMPTY_BUNDLE"
					) {
						return;
					}

					// TODO: pretty print warnings
				},
			} satisfies RollupOptions);

			const builder = await rollup(rollupOptions);

			const result = await builder.write({
				format: libraryType,
				sourcemap: false,
				dir: dir,
				entryFileNames: filename,
			});

			const outputs: BuildOutput[] = [];
			for (const _result of result.output) {
				const stats = await stat(join(cwd, "dist", _result.fileName));
				outputs.push({
					name: _result.fileName,
					path: join(cwd, "dist", _result.fileName),
					size: stats.size,
				});
			}

			return outputs;
		},
	});
}
