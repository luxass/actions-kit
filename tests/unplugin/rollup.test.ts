import { readFile } from "node:fs/promises";
import { join } from "node:path";
import typescriptPlugin from "@rollup/plugin-typescript";
import { rollup } from "rollup";
import { expect, it } from "vitest";
import { fromFileSystem, testdir } from "vitest-testdirs";
import ActionKitPlugin from "../../src/unplugin/rollup";

it("expect `actions-kit.d.ts` to be generated", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const bundle = await rollup({
    input: join(testdirPath, "index.ts"),
    external: ["@actions/core"],
    plugins: [
      typescriptPlugin({
        tsconfig: join(testdirPath, "tsconfig.json"),
      }),
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
      }),
    ],
  });

  const { output } = await bundle.generate({
    format: "cjs",
    sourcemap: false,
  });

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(output[0]).toBeDefined();
  expect(output[0].code).toBeDefined();

  expect(output[0].code).toMatchSnapshot();
  expect(dtsOutput).toMatchSnapshot();
});

it("expect `actions-kit.d.ts` to include `ACTION_INPUTS`", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const bundle = await rollup({
    input: join(testdirPath, "index.ts"),
    external: ["@actions/core"],
    plugins: [
      typescriptPlugin({
        tsconfig: join(testdirPath, "tsconfig.json"),
      }),
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
        injectInputs: true,
      }),
    ],
  });

  const { output } = await bundle.generate({
    format: "cjs",
    sourcemap: false,
  });

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(output[0]).toBeDefined();
  expect(output[0].code).toBeDefined();
  expect(output[0].code).toContain("ACTION_INPUTS");
  expect(dtsOutput).toContain("ACTION_INPUTS");

  expect(output[0].code).toMatchSnapshot();
  expect(dtsOutput).toMatchSnapshot();
});
