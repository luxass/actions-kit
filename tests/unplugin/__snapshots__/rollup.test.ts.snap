// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expect \`actions-kit.d.ts\` to be generated 1`] = `
"'use strict';

var core = require('@actions/core');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var core__namespace = /*#__PURE__*/_interopNamespaceDefault(core);

async function run() {
    const type = core__namespace.getInput("type");
    const whoToGreet = core__namespace.getInput("who-to-greet");
    const message = \`Hello \${whoToGreet}!\`;
    core__namespace.info(message);
    if (type === "error") {
        core__namespace.setFailed("This is an error!");
    }
    core__namespace.setOutput("message", message);
}
run().catch((err) => {
    console.error(err);
    core__namespace.setFailed(err.message);
});
"
`;

exports[`expect \`actions-kit.d.ts\` to be generated 2`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 1`] = `
"'use strict';

var core = require('@actions/core');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var core__namespace = /*#__PURE__*/_interopNamespaceDefault(core);

globalThis.ACTION_INPUTS = {
  "type": "type",
  "who-to-greet": "who-to-greet"
};
async function run() {
    const type = core__namespace.getInput("type");
    const whoToGreet = core__namespace.getInput("who-to-greet");
    const message = \`Hello \${whoToGreet}!\`;
    core__namespace.info(message);
    if (type === "error") {
        core__namespace.setFailed("This is an error!");
    }
    core__namespace.setOutput("message", message);
}
run().catch((err) => {
    console.error(err);
    core__namespace.setFailed(err.message);
});
"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 2`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        
declare global {
  export const ACTION_INPUTS: {
  "type": "type",
  "who-to-greet": "who-to-greet"
};
}

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;
